### ISSUE TITLE
Remember last window position

### ISSUE THREAD

**Issue by deluan** [fyne-io/fyne#554]
I'd like to have my app's windows re-open in the same spot as I left it last time.... Just a small nuisance to keep moving it every time I open it.

In macOS, this is the default behaviour, but Fyne apps don't behave like that. Not sure if this should be considered a bug or not.

I'm also aware that this is a desktop responsibility (positioning new windows), and that some Linux window managers have a "Remember" feature, but this is not consistent in all environments.

After briefly discussing this in Fyne's Slack channel, the proposal is to add a feature that would allow the app to request this behaviour, something like `window.RememberPosition()`. This would of course only work in environments that provide such capability

**Comment by e1fueg0** [comment]
> This is something that I personally think should be handled by the window manager and not the toolkit.

Unfortunately, Microsoft programmers think another way, so it would be very good if it would be implemented somehow.

Or, at least, developers would like to get something like `bool Maximized()` and `Maximize(bool)` in the `Window` interface, please. The rest of the remember code we can write by ourselves =)

**Comment by andydotxyz** [comment]
The challenge is balancing platform-specific expectations with operations that can be expected to work across all platforms.
As illustrated above this would have to be an extension for specific OS combinations which we have avoided up to this point.
We can implement the remember functionality without exposing `Maximized` - this feature would work on all platforms.

Why do you feel that Microsoft users expect software to start maximized when they did not ask it to?

**Comment by Bluebugs** [comment]
I would really dislike if Fyne was trying to do the remember instead of my windows manager on Linux. If it is expected that application do that on Windows, I guess it might make sense to have a specific logic for Windows to remember the window placement on Windows.

**Comment by e1fueg0** [comment]
> Why do you feel that Microsoft users expect software to start maximized when they did not ask it to?

I can't tell for all MS users, but I'm also a user, not just a developer. And, as a user, I do hate some apps don't remember their maximized state. Some people like small windows on their desktop. I like (and I'm sure I'm not alone in this =) to maximize windows, and I would like if an app at least has a setting to do this. That's why I'm asking for a _possibility_ to do this.

**Comment by e1fueg0** [comment]
In fact, I came here from #776, which is closed. Should I move my comments there?

**Comment by andydotxyz** [comment]
> And, as a user, I do hate some apps don't remember their maximized state.

Agreed, the windows should be remembered. If the OS will not do it then we could do that in the toolkit.
That is why this ticket exists but "add SetMaximized" tickets get closed. We want the behaviour not the APIs that can be abused.

**Comment by andydotxyz** [comment]
(Hopefully the new title is less ambiguous)

**Comment by e1fueg0** [comment]
@andydotxyz 
And please keep in mind that the main windows are usually being shown like this:
```golang
	w.Resize(fyne.NewSize(640, 460))
	w.ShowAndRun()
```
(that was a code from `main.go` of `fyne_demo`)
So if you plan to do automatic restoring, transparent to a developer, this may need to be overwhelmed.

**Comment by stuartmscott** [comment]
`SetFallbackSize(fyne.Size)`?  If there is a user preference/previous setting use that, else fallback to the given size

**Comment by e1fueg0** [comment]
@andydotxyz 
I'd vote for the name `SetFallbackSize(fyne.Size)` proposed by @stuartmscott.

> I think it would be unwise to start ignoring APIs that have an expected outcome

Completely agree with you @andydotxyz, and I would have wanted to propose to add a field like `app.StoreMainWindowPosition`, but I suppose every Linux developer would turn it off, so the mechanism would never work on Windows =)

**Comment by Goltanju** [comment]
Just want to add another use case for this sort of programmatic behavior.

I have an application that is usually just one main window. The user can launch a slideshow however, which opens a new window. I'd like that new window to open maximized. Fyne currently won't let me do that (I have to hack the repo myself).

The "remembering" thing is something I also want, but how would remembering work with additional windows? Would I need to have a unique identifier for every window remember-me-type I want? That could work. Then the first time the user opens a slideshow it wouldn't be maximized, but it would be stay maximized if they closed as such. Not exactly what I want, but "okay".

I do know about the Fullscreen attribute. I don't want that in this case, as it hides absolutely everything, taskbar, other program's widgets, etc. I just want a maximized window, lol.

I also get that Fyne is trying to stay platform agnostic as much as possible, but I mean, a balance between that and providing long known features users would expect on a given platform... This particular hyper-focus on "no features unless all platforms support it" can make working with Fyne quite frustrating at times.

Anyway, back to hacking my fork...

**Comment by andydotxyz** [comment]
Yea it should work with additional windows.

Regarding your "edit: it would be nice if..." - please open a feature request to ask for new functionality.

**Comment by drunlade** [comment]
Lacking this feature is actually a bit of an issue. What I don't understand is people saying "It shouldn't be done because I don't want to use it", it's an optional feature, so don't use it? But for people who **need** it, then yeah, it should be implemented, even if it's not quite platform agnostic. The feedback given on our application was "It's frustrating that the window position and size always resets when loading a new instance, please fix this", but since the tooling is not there, we literally can't fix this.

**Comment by Jacalz** [comment]
In my case, it is simply from a user perspective that I don't want to use it; I wouldn't want a developer forcing this behaviour on me for any apps that I use because my experience is simply that it often leads to issues with different monitor configurations. I just want my apps to always pop up in the center of the monitor that I started them on. LibreOffice drives me mad sometimes when it tries to be smart about it. However, I do understand that some people may wish to use it (mostly on Windows if I understand correctly? I believe some window managers on Linux already has this option). If this is to be added, I would want the user to be able to disable it from something like fyne_settings etc.

**Comment by andydotxyz** [comment]
> Lacking this feature is actually a bit of an issue. 

Yes. If it was not then the ticket would have been closed already. You can see also it is aiming for the next major release (see "Milestone").

> What I don't understand is people saying "It shouldn't be done because I don't want to use it"

I don't really think that is what is being said above, it is a discussion about how/when it should be handled.

> it's an optional feature, so don't use it?

That is where things get complicated. I don't think this can be optional - because by turning it off you would expect it to _not_ happen but on some platforms it already does. Therefore, for greater consistency, the only logical approach is to always do it. The question is much more "how" than "should we".

> but since the tooling is not there, we literally can't fix this

In case it is helpful your statement is not technically true. Since 2.5 you could access the window handle and persist the position for yourself, and then load it again. However it's a lot of complex code which really belongs in the library so the better solution is for this ticket to get resolved.

**Comment by andydotxyz** [comment]
> If this is to be added, I would want the user to be able to disable it from something like fyne_settings etc.

Yes, that makes sense. If it must be optional it should be up to the user, not up to the developer, to decide how windows are placed on the desktop.

**Comment by dweymouth** [comment]
> The feedback given on our application was "It's frustrating that the window position and size always resets when loading a new instance, please fix this", but since the tooling is not there, we literally can't fix this.

Just to point out, it **is possible** to remember the window size with the current APIs. You can use window.Canvas().Size() to query the window size when your app exits, use the preferences API to save it, and then use window.Resize(...) to start your app at the last window size in your main.go. I've had this in my own Fyne app since its initial release.

Window positioning is what's currently not possible with the Fyne APIs (at least without the window.RunNative API and custom native code), and to be honest I do kind-of feel like it *should* be on the OS / window manager to position windows appropriately, but it's a reasonable feature request for Fyne nonetheless, since many WMs don't actually do this well.

**Comment by drunlade** [comment]
@andydotxyz I got the wrong impression then, I read it's from 2019 with what seemed like no real progress, though I admit I perhaps read some of the comments with a more negative tone than was intended, so my mistake. I was in a frustrated mood a the time and it coloured my response.

> see "Milestone"

👀  Ok yes, I did entirely miss the milestone.

> it's an optional feature, so don't use it?
>> That is where things get complicated. I don't think this can be optional - because by turning it off you would expect it to not >> happen but on some platforms it already does. Therefore, for greater consistency, the only logical approach is to always do >> it. The question is much more "how" than "should we".

This doesn't make a lot of sense to me. Devs have been wanting the ability to do this for five years now, and we're saying the main thing holding it back is "some windows managers already do this"? The most logical approach in my mind is to get the feature implemented and make it easy to toggle, it's not hard for a dev to add an "is the OS linux" check and disable it early if that's what they want. Let devs make up their own minds?

> In case it is helpful your statement is not technically true. Since 2.5 you could access the window handle and persist the
> position for yourself, and then load it again. However it's a lot of complex code which really belongs in the library so the better solution is for this ticket to get resolved.

I wasn't aware of that no, I agree though that messing with the underlying handle is a lot less desirable than having a `.SetPosition(x,y,screen)` type method.

**Comment by dweymouth** [comment]
> That is where things get complicated. I don't think this can be optional - because by turning it off you would expect it to not happen but on some platforms it already does. Therefore, for greater consistency, the only logical approach is to always do it. The question is much more "how" than "should we".

To add to Jacob's point that it should be optional as a user setting, I'd argue that it should - at least in some cases - be developer-optional as well. For example, an app that is intending to run in "portable"/"leave no trace" mode should not have Fyne writing anything persistent to the system with no means to disable it.

**Comment by andydotxyz** [comment]
>  The most logical approach in my mind is to get the feature implemented and make it easy to toggle

We are in agreement here.

> it's not hard for a dev to add an "is the OS linux" check and disable it early

Our aim is to eliminate OS specific code where possible, not add to it.

> Let devs make up their own minds?

As above I don't think the developer knows the user intent best - the users pick the OS or window manager and should be left in control of window positioning.

> a lot less desirable than having a .SetPosition(x,y,screen) type method.

That is a whole other discussion - see #1155

**Comment by andydotxyz** [comment]
> For example, an app that is intending to run in "portable"/"leave no trace" mode should not have Fyne writing anything persistent to the system with no means to disable it.

If we want to provide the intent of "leave no trace" then we should provide that as an API. I don't think developers would be well served to have an "off" switch for each individual area that could possibly write a preference and turn every one of them off specifically. It comes back to the behaviour based API design. There are already many widgets that write a preference value.

**Comment by drunlade** [comment]
> Let devs make up their own minds?
>> As above I don't think the developer knows the user intent best - the users pick the OS or window manager and should be left in control of window positioning.

I understand you have a goal, a toolkit that results in very consistent results across all platforms, but "_the users pick the OS or window manager and should be left in control of window positioning_" really doesn't apply when you look outside of home users. Take corporate software, a developer of that sort of software often needs to be able to control how the software behaves in every single way because it's going to be distributed on thousands of near-identical systems operated by people with zero technical knowledge. It has to work in the way the devs designed, not up to the whims of the OS or some Fyne default.

Of course, this becomes an issue when the OS explicitly **blocks** this functionality, which I know is the case for Wayland, but I don't think that's a good argument for saying the feature therefor should not exist (or should be heavily restricted), preventing the Windows exclusive devs from creating functionality they happen to require. "We won't let you do X because you can't do it on Y platform" just feels unhelpful to all the devs who need the feature on the platforms that do support it.

I can see multi-window etc is covered already in #1155 so I won't get into that example.

> Our aim is to eliminate OS specific code where possible, not add to it.

I agree with this goal, I really do, but less so in cases where it directly hinders (or outright blocks) the ability for a dev to create what they want/need. I think that when a lib is mobile-only I would agree completely, absolute consistency is basically a requirement, but when desktops are in the mix, flexibility is often king. I am well aware it's a hard balance to strike, and my opinion is naturally influenced by the fact I happen to need this feature and am only targeting Windows platforms.

**Comment by andydotxyz** [comment]
> I don't think that's a good argument for saying the feature therefor should not exist (or should be heavily restricted), preventing the Windows exclusive devs from creating functionality they happen to require. "We won't let you do X because you can't do it on Y platform" just feels unhelpful

Please stop implying that we/I think this feature should not exist. That is not true and is not helpful. We have agreed windows should remember position.

> a developer of that sort of software often needs to be able to control how the software behaves in every single way because it's going to be distributed on thousands of near-identical systems

In that case the configuration above can indeed be set to override the defaults.

> I think that when a lib is mobile-only I would agree completely, absolute consistency is basically a requirement, but when desktops are in the mix, flexibility is often king.

I don't agree that mobile apps should all be the same but desktop apps should all behave differently - but we do know the importance of configuration. That is what has been outlined above. Adding this feature *and* making it possible to be turned off. I don't know why you think we are not listening or don't want to implement it.

**Comment by davidmichaelkarr** [comment]
I only just started working with Fyne, so I'm coming into this late.

I implemented this on Windows by saving the current size and position in preferences, and for the troublesome position, I used reflection to call Viewport.SetPos, which appears to be working.  Is the argument against allowing easier access to this that it simply won't work (ignored?) on some platforms? I think I would agree with some posters that this isn't a good argument for not allowing it.

**Comment by andydotxyz** [comment]
I may be confused - but this ticket is about remembering window state and as far as I can see we have agreed that it should be done.
So what's the chat about it not being allowed?

**Comment by dweymouth** [comment]
There was conversation about a `SetPosition` API which is a slightly different request than remembering position. (and tracked in https://github.com/fyne-io/fyne/issues/1155) Though you did mention that it would be a good API for the new `desktop.Window` interface introduced in https://github.com/fyne-io/fyne/pull/6153

Edit: forgot that the RequestPosition API was added for 2.8 in #6184. Closed out issue 1155
